addons/delivery/models/stock_picking.py

```diff
@@ -13,21 +14,31 @@ class StockQuantPackage(models.Model):
 
     @api.depends('quant_ids')
     def _compute_weight(self):
+        if self.env.context.get('picking_id'):
+            package_weights = defaultdict(float)
+            # Ordering by qty_done prevents the default ordering by groupby fields that can inject multiple Left Joins in the resulting query.
+            res_groups = self.env['stock.move.line'].read_group(
+                [('result_package_id', 'in', self.ids), ('product_id', '!=', False), ('picking_id', '=', self.env.context['picking_id'])],
+                ['id:count'],
+                ['result_package_id', 'product_id', 'product_uom_id', 'qty_done'],
+                lazy=False, orderby='qty_done asc'
```

Comment:
ordering by id isn't event more efficient, as qty_done is not indexed?

Reply:
@Whenrow The issue is that the order key's column has to appear either in the `Group By` clause or in an aggregate function. I cannot use an indexed many2one in the order as `_generate_order_by` will use the model's `_order` attribute, which is something I want to avoid. I cannot add the id column inside the `Group By` clause as that would remove the purpose of the group by. I could potentially order by `min(id)` or `count(id)` but I'm not sure that it will be more efficient nor do I think that the read_group's method allows me to do that.

Reply:
Not min or count, use only id:
https://github.com/odoo/odoo/blob/0e11f4926be1800fc2b58feedcd81df24720ec19/addons/stock/models/product.py#L171-L173

Reply:
Yes but postgres needs to know how to produce a single value from all the aggregated rows for the `id` column, either by using it in the `GROUP BY`, ensuring its uniqueness, or by using an aggregate function to produce a single value from a group of ids. In the example above it's not really the `id` column that is used as a sorting key. If you look at one of the produced query (I simplified the where clause to enhance the readability).

```sql
SELECT min("stock_move".id) AS id, 
       count("stock_move".id) AS "product_id_count" , 
       sum("stock_move"."product_qty") AS "product_qty",
       "stock_move"."product_id" as "product_id" 
FROM "stock_move" 
LEFT JOIN "stock_location" AS "stock_move__location_dest_id" 
       ON ("stock_move"."location_dest_id" = "stock_move__location_dest_id"."id") 
LEFT JOIN "stock_location" AS "stock_move__location_id" 
       ON ("stock_move"."location_id" = "stock_move__location_id"."id")
WHERE {where clause}
GROUP BY "stock_move"."product_id"
ORDER BY "id";
```

Here the `id` column used in `ORDER BY` does not refer to the `id` column of stock_move, instead it refers to the id-renamed column in the SELECT clause, i.e. `min("stock_move".id)`. You can see that in the query plan here: https://explain.dalibo.com/plan/e3gda18a37df1542 The sort key of the first SORT node is `min("stock_move".id)` and not simply `id`. I could have used `orderby="id"` in the read_group for consistency across modules but performance-wise it should be the same as ordering by `qty_done`